java/adapter/jdbc/src/test/java/org/apache/arrow/adapter/jdbc/ResultSetUtility.java

```diff
@@ -0,0 +1,1617 @@
+    public boolean wasNull() throws SQLException {
+      return wasNull;
+    }
+
+    public static class Builder {
+      private final ArrayList<MockRow> rows;
+      private ArrayList<MockDataElement> bufferedElements;
+      private ResultSetMetaData metadata;
+
+      Builder() {
+        this.rows = new ArrayList<>();
+        this.bufferedElements = new ArrayList<>();
+      }
+
+      public Builder finishRow() {
+        rows.add(new MockRow(this.bufferedElements));
+        this.bufferedElements = new ArrayList<>();
+        return this;
+      }
+
+      public Builder addDataElement(MockDataElement element) {
+        this.bufferedElements.add(element);
+        return this;
+      }
+
+      public Builder addDataElement(String str) {
+        return this.addDataElement(new MockDataElement(str));
+      }
+
+      public Builder addDataElement(Object val, int sqlType) {
+        return this.addDataElement(new MockDataElement(val, sqlType));
+      }
+
+      public Builder setMetaData(ResultSetMetaData metaData) {
+        this.metadata = metaData;
+        return this;
+      }
+
+      public MockResultSet build() throws SQLException {
+        if (this.metadata == null) {
+          return new MockResultSet(this.rows);
+        }
+        return new MockResultSet(this.rows, this.metadata);
+      }
+    }
+  }
+
+  public static class MockResultSetMetaData extends ThrowingResultSetMetaData {
+    private ArrayList<MockColumnMetaData> columns;
+
+    public MockResultSetMetaData(ArrayList<MockColumnMetaData> columns) {
+      this.columns = columns;
+    }
+
+    @Override
+    public int getColumnCount() throws SQLException {
+      return columns.size();
+    }
+
+    @Override
+    public String getColumnLabel(int column) throws SQLException {
+      return columns.get(column - 1).getLabel();
+    }
+
+    @Override
+    public String getColumnName(int column) throws SQLException {
+      return columns.get(column - 1).getName();
+    }
+
+    @Override
+    public int getColumnType(int column) throws SQLException {
+      return columns.get(column - 1).getType();
+    }
+
+    @Override
+    public int getPrecision(int column) throws SQLException {
+      return columns.get(column - 1).getPrecision();
+    }
+
+    @Override
+    public int getScale(int column) throws SQLException {
+      return columns.get(column - 1).getScale();
+    }
+
+    @Override
+    public int isNullable(int column) throws SQLException {
+      return columns.get(column - 1).isNullable();
+    }
+
+    public static MockResultSetMetaData fromRows(ArrayList<MockRow> rows) throws SQLException {
+
+      if (rows.size() == 0) {
+        throw new SQLException("Unable to dynamically generate ResultSetMetaData because row count is zero!");
+      }
+      MockRow firstRow = rows.get(0);
+      if (firstRow.dataElements.size() == 0) {
+        throw new SQLException("Unable to dynamically generate ResultSetMetaData because column count is zero!");
```

Comment:
Is it not valid to have a zero-column dataset?

Reply:
If it is, it's not possible to dynamically construct a MockResultSetMetaData from a zero-column data set.  I've added a comment to clarify.